Writing MatlabOpaque objects to MAT-files

foreverallama opened the pull request:
Part of a series of commits to support write operations for `MatlabOpaque` objects. For now I just rearranged the methods to make it easy for me to update the code. 

I'll first add to `MAT_subsys.jl` before updating `MAT_HDF` and `MAT_v5`.

foreverallama commented:
Most of the code's ready. It's not working yet I'll debug soon and in the meantime add some sample tests

matthijscox commented on src/MAT_subsys.jl:
```diff
@@ -394,4 +428,300 @@ function load_mcos_object(metadata::Array{UInt32}, type_name::String, subsys::Su
+    push!(fwrap_data, Any[])
+    append!(fwrap_data, subsys.prop_vals_saved)
+
+    empty_struct = MatlabStructArray([]) #! FIXME
```
Reading the code, this might not yet be possible. I always estimate the dimensions from the first column, but if there are no columns... Might be we have to explicitly add the dimensions to the MatlabStructArray type, so that it also works without any columns.

Does the empty struct array have any special features inside the HDF5 format?

foreverallama replied:
> easy way to load an example fully empty (no columns, 1x0 matrices) struct array type

Sorry, yeah that's a good point. I'll check this out first

foreverallama commented on src/MAT_HDF5.jl:
```diff
@@ -682,6 +689,20 @@ end
 
 # Write a struct from arrays of keys and values
 function m_write(mfile::MatlabHDF5File, parent::HDF5Parent, name::String, k::Vector{String}, v::Vector)
+    if length(k) == 0
```
I added the ability to write empty dicts as 0x0 structs over here. The subsystem data works with 0x0 structs as well in MATLAB. 

This eliminates the need for any new struct markers for now.

matthijscox replied:
Alright. Are 0x0 structs also read as empty dicts?

foreverallama replied:
Yup, all of 1x0, 0x1 and 0x0 are read as empty dicts already. But I hardcoded 0x0 for empty dict during write.

matthijscox replied:
It's a rather special corner case. I'm okay with a new type that we keep internal for the writing, and then we can always consider to refactor later:
```julia
struct EmptyStruct
   dims::Vector{Int}
end
```

foreverallama commented:
* Added support for writing object arrays
* Added simple test for basic objects, handle class objects, object arrays and nested objects
* Empty Dict is now written in as a 1x1 struct with no fields (fixes #214 )
* New type `EmptyStruct` to write special zero-dimensional structs (for internal use)

The objects being written are loading successfully in MATLAB (2023b). With this I think the base for writing objects is covered. We can now write methods for converting from say `datetime` etc. to `MatlabOpaque`. A point to note is that subsystem data is being written in version 4. We can also write in v2 instead, though I think v4 has been the standard for a while now.  

There's a special case of zero-dimensional objects, but again its an edge case that can be tackled if required.

I also opened #213 to fix #212 , it would be easier to merge that first and then fix conflicts here.

matthijscox commented:
Do the tests now also cover the empty dict read/write case? If not, it would be good to add a test for that, so we don't accidentally break it again in the future.

matthijscox commented on test/write.jl:
```diff
@@ -193,4 +193,111 @@ nt_read = matread(tmpfile)["nt"]
+        @test read_var["var1"] === read_var["var2"]  # same object
+    end
+end
+
```
I don't see tests for the special FileWrapper cases? Is that intentional?

foreverallama replied:
Which one do you mean?

matthijscox commented on src/MAT_HDF5.jl:
```diff
@@ -541,14 +543,19 @@ function m_write(mfile::MatlabHDF5File, parent::HDF5Parent, name::String, c::Abs
         write_dataset(cset, ctype, refs)
-        write_attribute(cset, name_type_attr_matlab, "cell")
+        if object_decode == UInt32(3)
+            write_attribute(cset, object_decode_attr_matlab, object_decode)
```
is this (FileWrapper) case tested?

matthijscox commented:
Alright, everything looks okay to me. Do you have the power to merge already (since you are a contributor now) or do I have to make you an official maintainer for that?

matthijscox commented:
I'm only a member of JuliaIO I see, so cannot add you as maintainer unfortunately. I'll just have to merge PRs I suppose

foreverallama commented:
No issues, you can also close #214
